tests/babel_ts/test.ts

```diff
@@ -0,0 +1,17 @@
+function add(a: number, b: number): number;
+function add(a: string, b: string): string;
+function add(a: any, b: any) {
+  return a + b;
+}
+
+class Foo1 {
+  private foo() {}
+}
+
+// abstract class Foo2 {
+//   abstract foo ();
+// }
+
+// declare class Foo3 {
+//   foo();
+// }
```

Comment:
Still in WIP?

Reply:
Yep, I will try to fix them after finishing my own job tonight.

Reply:
@evilebottnawi Sorry, there are some bad things happened in my family, so I will not have time to do this in few days maybe. Hope someone else can continue the work,

Reply:
@JounQin sympathize, what problems remained?

Reply:
@evilebottnawi It doesn't support `TSDeclareMethod` type as @
thorn0 's description. And includeJsx param can be moved into extraPlugins.

Reply:
It's already late in China, go take care you family.

Reply:
If we enable this parser in all the tests where `typescript` is enabled, we'll see what's left to do here. BTW I wonder why not all the checks ran for the latest changes in this PR.

Reply:
@thorn0 maybe you can help with this PR?

Reply:
@evilebottnawi I'd like to, but I can't promise I'll have time for it in the near future.

Reply:
@thorn0 I don't think we should support all features in only one PR personally.

Reply:
@JounQin Things like issues with comments and blank lines can be solved later. E.g. I just found out that the tests in `tests/comments` aren't enabled for the `typescript` parser because there are still lots of inconsistencies there. But as for the core syntax (not comments), let's cover it from the beginning, shall we? TS 3.7 is about to be out, and for now, parsing with Babel seems to be the only solution Prettier 1.19 is going to offer to those who want to use nullish coalescing and optional chaining.

Reply:
Of course it is great to support them by once, and whoever wants these features can continue this PR, or they can only wait.

Reply:
There is no point in merging while the printer doesn't understand what the parser produces. Unless we mark it as 'experimental' or something, and still I wouldn't see the point.

Reply:
We can merge this, but don't provide publicly as we do for handlebars https://github.com/prettier/prettier/blob/master/src/language-handlebars/index.js#L9